hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java

```diff
@@ -48,8 +48,12 @@ public class ContainerHealthStatus {
     int repFactor = container.getReplicationConfig().getRequiredNodes();
     this.healthyReplicas = healthyReplicas
         .stream()
-        .filter(r -> !r.getState()
-            .equals((ContainerReplicaProto.State.UNHEALTHY)))
+        // Filter unhealthy replicas and
+        // replicas belonging to out-of-service nodes.
+        .filter(r ->
+            (!r.getDatanodeDetails().isDecommissioned() &&
+             !r.getDatanodeDetails().isMaintenance() &&
```

Comment:
It is "OK" for a maintenance replica to be offline. The definition of maintenance is that one or two replicas out of 3 can be offline and the container is still considered healthy, so I am not sure if it is correct to just assume a maintenance copy is offline. If there was a scenario where all 3 of the replicas of a container were in maintenance, with no other available copies, then that would be under-replicated.

This leads to another question - what if a container has 3 replicas from alive nodes:

IN_SERVICE
IN_SERVICE
IN_MAINTENANCE

Then the IN_MAINTENANCE node gets shutdown, so there are only the 2 IN_SERVICE nodes left. This is not considered under-replicated and RM will take no action on this container, as it believes that the IN_MAINTENANCE copy will return.

SCM keeps track of replicas that are for Maintenance nodes and counts them as still being available so it can make decisions on this sort of thing.

The concern I have is that Recon shows different counts for under-replicated than the RM report and it can cause confusion to users.

Reply:
Like you said, for SCM, healthy replicas are also the ones that are offline [getHealthyReplicaCount()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L193-L196).

When the SCM checks for over-replication or under-replication, it gets only the available replicas [getAvailableReplicas()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L215-L221) based mostly on the `healthyReplicaCount` field.

The `healthyReplicaCount` doesn't count replicas from nodes in decommission or maintenance [countReplicas()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L171).

```
this.healthyReplicas = healthyReplicas
```

I know that it says `healthyReplicas` but it should hold just the available replicas in order to be consistent with the SCM.

Reply:
As far as I understand, a node doesn’t go offline until its replicas have been copied to another node. While ENTERING_MAINTENANCE or DECOMMISSIONING container replicas are added or removed as needed to maintain proper replication. The container will be under-replicated until copies have been made and the node successfully becomes offline.

Once that is done, the container is correctly replicated, has 3 healthy and available replicas and 1 offline. SCM doesn’t report any under-replicated or over-replicated containers but Recon 

- for master, counts 1 over-replicated because it sees 4 replicas (no distinction between online - offline).
- for this patch, 0 count.

When the offline datanode is stopped, SCM doesn’t count unhealthy containers and

- for master, Recon no longer counts 1 over-replicated container.
- for this patch, no change in Recon.